### PR TITLE
Declare permissions

### LINKED ISSUE #15462: Workflows are missing permissions requests
https://github.com/check-spelling-sandbox/codeql/actions/runs/7699091660/workflow

https://github.com/check-spelling-sandbox/codeql/actions/runs/7699091660/job/20979906681#step:19:55

```
Post job cleanup.
Warning: Debugging artifacts are unavailable since the 'init' Action failed before it could produce any.
RequestError [HttpError]: Resource not accessible by integration
    at /home/runner/work/_actions/github/codeql-action/v2/node_modules/@octokit/request/dist-node/index.js:86:21
    at processTicksAndRejections (node:internal/process/task_queues:96:5)
    at async requestWithGraphqlErrorHandling (/home/runner/work/_actions/github/codeql-action/v2/node_modules/@octokit/plugin-retry/dist-node/index.js:71:20)
    at async Job.doExecute (/home/runner/work/_actions/github/codeql-action/v2/node_modules/bottleneck/light.js:405:18) {
  status: 403,
  response: {
    url: 'https://api.github.com/repos/check-spelling-sandbox/codeql/code-scanning/analysis/status',
    status: 403,
    headers: {
      'access-control-allow-origin': '*',
      'access-control-expose-headers': 'ETag, Link, Location, Retry-After, X-GitHub-OTP, X-RateLimit-Limit, X-RateLimit-Remaining, X-RateLimit-Used, X-RateLimit-Resource, X-RateLimit-Reset, X-OAuth-Scopes, X-Accepted-OAuth-Scopes, X-Poll-Interval, X-GitHub-Media-Type, X-GitHub-SSO, X-GitHub-Request-Id, Deprecation, Sunset',
      connection: 'close',
      'content-encoding': 'gzip',
      'content-security-policy': "default-src 'none'",
      'content-type': 'application/json; charset=utf-8',
      date: 'Mon, 29 Jan 2024 16:16:02 GMT',
      'referrer-policy': 'origin-when-cross-origin, strict-origin-when-cross-origin',
      server: 'GitHub.com',
      'strict-transport-security': 'max-age=31536000; includeSubdomains; preload',
      'transfer-encoding': 'chunked',
      vary: 'Accept-Encoding, Accept, X-Requested-With',
      'x-content-type-options': 'nosniff',
      'x-frame-options': 'deny',
      'x-github-api-version-selected': '2022-11-28',
      'x-github-media-type': 'github.v3; format=json',
      'x-github-request-id': '94E3:79CB:3878D:7311B:65B7CF42',
      'x-ratelimit-limit': '1000',
      'x-ratelimit-remaining': '961',
      'x-ratelimit-reset': '1706548556',
      'x-ratelimit-resource': 'core',
      'x-ratelimit-used': '39',
      'x-xss-protection': '0'
    },
    data: {
      message: 'Resource not accessible by integration',
      documentation_url: 'https://docs.github.com/rest'
    }
  },
  request: {
    method: 'PUT',
    url: 'https://api.github.com/repos/check-spelling-sandbox/codeql/code-scanning/analysis/status',
    headers: {
      accept: 'application/vnd.github.v3+json',
      'user-agent': 'CodeQL-Action/2.23.2 octokit-core.js/3.6.0 Node.js/16.20.2 (linux; x64)',
      authorization: 'token [REDACTED]',
      'content-type': 'application/json; charset=utf-8'
    },
    body: '{"action_name":"init-post","action_oid":"unknown","action_ref":"v2","action_started_at":"2024-01-29T16:16:02.392Z","action_version":"2.23.2","analysis_key":".github/workflows/ql-for-ql-dataset_measure.yml:measure","commit_oid":"aeae208dc3291109d6c798179bb8944961348823","job_name":"measure","job_run_uuid":"75681a8a-17f0-4c74-b850-172cffab9a66","ref":"refs/heads/main","runner_available_disk_space_bytes":31716970496,"runner_os":"Linux","runner_total_disk_space_bytes":89297309696,"started_at":"2024-01-29T16:16:01.614Z","status":"success","testing_environment":"","workflow_name":"Collect database stats for QL for QL","workflow_run_attempt":1,"workflow_run_id":7699091660,"completed_at":"2024-01-29T16:16:02.416Z","matrix_vars":"{\\n  \\"repo\\": \\"github/codeql\\"\\n}","runner_arch":"X64","runner_image_version":"20240126.1.0","job_status":"JOB_STATUS_UNKNOWN"}',
    request: { agent: [Agent], hook: [Function: bound bound register] }
  }
}
Error: Resource not accessible by integration
```

I presume that it needs:

```
permissions:
  security-events: write
```

or similar, but this api isn't documented in https://docs.github.com/en/rest/authentication/permissions-required-for-github-apps?apiVersion=2022-11-28 so I have absolutely no idea.

### PR DESCRIPTION
**Pull request by jsoref**
Repositories can be configured with Default access (restricted) https://docs.github.com/en/actions/security-guides/automatic-token-authentication#permissions-for-the-github_token

Best practice says that workflows should declare the minimal permissions they require. Without declaring permissions, paranoid forks fail miserably.

closes #15462

### CODE REVIEW

File: csharp/actions/create-extractor-pack/action.yml
Comment (angelapwen) [inline csharp/actions/create-extractor-pack/action.yml:6]:
```suggestion

# permissions:
#   security-events: read

```

**Comment by jsoref** [comment]
To make the Check change note workflow happy, please add `label:no-change-note-required`

File: csharp/actions/create-extractor-pack/action.yml
Comment (angelapwen) [inline csharp/actions/create-extractor-pack/action.yml:6]: Nit: could you remove this?

**Review by angelapwen — APPROVED** [review]
I looked at your fork of the repo, which has identical permissions and works 👍 thank you for the contribution!

File: csharp/actions/create-extractor-pack/action.yml
Comment (jsoref) [reply csharp/actions/create-extractor-pack/action.yml:6]: The reason I put it there is that it's really helpful to be able to look at actions to understand what they require. (And I _think_ that's what that action requires.)

But, I'll remove it.

Comment (angelapwen) [reply csharp/actions/create-extractor-pack/action.yml]: Oh, I see! That's a good idea. Do other Actions use this method to declare what they require? I don't see that we have it in the CodeQL Action, for instance. If it's not a known standard, how about:

```
# Requires `security-events: read` permissions` 
```
or something along those lines?

Comment (jsoref) [reply csharp/actions/create-extractor-pack/action.yml]: The _better_ behaved repositories w/ a single top level action.yml declare their required permissions in the README.md at the top level.

For example [`actions/stale`](https://github.com/actions/stale?tab=readme-ov-file#recommended-permissions) has a section for it (standard disclaimer: I hate the action).

For repositories that have multiple actions, it's harder -- ideally they'd use README.md files to document how to use the actions, but creating such a file seemed a bit much.

Comment (angelapwen) [reply csharp/actions/create-extractor-pack/action.yml]: That makes sense — thanks for sharing. I'll keep it in mind as it relates to the CodeQL Action, and re-approving here!

**Review by angelapwen — APPROVED** [review]

File: .github/workflows/check-implicit-this.yml
Comment (angelapwen) [inline .github/workflows/check-implicit-this.yml:14]: I've been reminded to check that all of these permissions are necessary. Undoubtedly the `security-events:write` one for the QL-for-QL tests workflow is, as we saw elsewhere. 

I tried this one in particular out on my own fork (where the current commit is `main` and where I've set the default token permissions to read-only) and it looks like [this one passes without adding the extra permission](https://github.com/angelapwen/codeql/actions/workflows/check-implicit-this.yml). Is this because `read` is allowed with the least permissive setting?

These are the options I have:
Screenshot 2024-02-01 at 11 14 53 AM

It would be nice to confirm that all the added permissions are necessary.

File: .github/workflows/labeler.yml
Comment (angelapwen) [inline .github/workflows/labeler.yml:8]: This one, for example, already has a permissions block in the single job below. I do understand having it at a top-level for readability, but it'd be great if we could remove the other under `triage`.

**Review by angelapwen — CHANGES_REQUESTED** [review]
Blocking merge until the other comments are addressed 😄

File: .github/workflows/check-implicit-this.yml
Comment (jsoref) [reply .github/workflows/check-implicit-this.yml:14]: While this looks like a grant, it's actually a restriction. By default you get some additional permissions:
```
  Contents: read
  Metadata: read
  Packages: read
```

By specifying `contents: read`, it's blocking `packages: read`.

File: .github/workflows/labeler.yml
Comment (jsoref) [reply .github/workflows/labeler.yml:8]: Either works.

I think I do prefer using a top-level permission default and then only overriding in workflows with multiple jobs that need distinct permissions as opposed to having workflows with single jobs where one has to look into the job for the permission.

File: .github/workflows/check-implicit-this.yml
Comment (jsoref) [reply .github/workflows/check-implicit-this.yml:14]: `contents: read` is theoretically needed for any workflow that uses `actions/checkout` (it might not realistically fail in public forks of a repository, but it would absolutely fail if someone made a private repository with the same contents).

Comment (jsoref) [reply .github/workflows/check-implicit-this.yml:14]: I should note that I do things like that regularly when I'm testing repositories looking for vulnerabilities.

(It isn't my day job, but I do it often enough. It's definitely considered best practice. Admittedly, w/o a GHES license w/ CodeQL it'd be a bit of a PITA here, but it's technically possible to manage if you used a second public repository with a PAT for the codeql bits. But the general idea here is to show best practices so that if people copy+paste they get best practices.)

File: .github/workflows/csv-coverage-update.yml
Comment (aeisenberg) [inline .github/workflows/csv-coverage-update.yml:10]: Is this right? Should it be this?

```suggestion
  pull-requests: write
```

`scripts/library-coverage/create-pr.py` will open a new PR. I don't see anywhere that it needs to upload security-related things. Might be mistaken, though.

@jsoref can you run this in your fork to see if it works?

File: .github/workflows/ql-for-ql-tests.yml
Comment (aeisenberg) [inline .github/workflows/ql-for-ql-tests.yml:22]: Is `security-events` necessary at all? I don't see it uploading SARIF.

```suggestion
```

File: .github/workflows/query-list.yml
Comment (aeisenberg) [inline .github/workflows/query-list.yml:18]: Necessary?

```suggestion
```

File: .github/workflows/ruby-dataset-measure.yml
Comment (aeisenberg) [inline .github/workflows/ruby-dataset-measure.yml:22]: ?

```suggestion
```

File: .github/workflows/ruby-qltest.yml
Comment (aeisenberg) [inline .github/workflows/ruby-qltest.yml:34]: ?

```suggestion
```

File: .github/workflows/swift.yml
Comment (aeisenberg) [inline .github/workflows/swift.yml:38]: ?

```suggestion
```

File: .github/workflows/validate-change-notes.yml
Comment (aeisenberg) [inline .github/workflows/validate-change-notes.yml:20]: ?

```suggestion
```

**Review by aeisenberg — CHANGES_REQUESTED** [review]
Thanks for contributing this. I like that this narrows the permissions of all of our tokens.

I think `.github/workflows/csv-coverage-update.yml` is broken. I have a few other suggestions that will allow you to narrow the permissions. `security-events` is only required if we are reading or writing SARIF to or from code scanning.

I do have a concern that these jobs haven't been run. Are you able to trigger them in your fork to make sure all permissions are correct?

What we do in other repos and can do here (but best to wait for a followup PR) is to add a chunk that ensures the workflow file is run whenever the workflow file itself is modified.

eg-

```
  pull_request:
    paths:
      - '.github/workflows/csv-coverage-timeseries.yml'
```

(and similar for all other workflow files)

**Comment by jsoref** [comment]
@aeisenberg: the design of these workflows is really painful.

* Many of them run only on specific branches.
* Many of them only run if specific files change.
* Many of them really really only work if run from a PR. -- I tried to add workflow_dispatch to test things and things exploded spectacularly: https://github.com/check-spelling-sandbox/codeql/actions/runs/7818256643/job/21328047264

File: .github/workflows/check-implicit-this.yml
Comment (jsoref) [reply .github/workflows/check-implicit-this.yml:14]: https://github.com/check-spelling-sandbox/codeql/actions/runs/7820458947/job/21335230118

```
GITHUB_TOKEN Permissions
  Contents: read
  Metadata: read
```

```
Warning: CodeQL Action v2 will be deprecated on December 5th, 2024. Please update all occurrences of the CodeQL Action in your workflow files to v3. For more information, see https://github.blog/changelog/2024-01-12-code-scanning-deprecation-of-codeql-action-v2/
RequestError [HttpError]: Resource not accessible by integration
    at /home/runner/work/_actions/github/codeql-action/v2/node_modules/@octokit/request/dist-node/index.js:86:21
    at processTicksAndRejections (node:internal/process/task_queues:96:5)
    at async requestWithGraphqlErrorHandling (/home/runner/work/_actions/github/codeql-action/v2/node_modules/@octokit/plugin-retry/dist-node/index.js:71:20)
    at async Job.doExecute (/home/runner/work/_actions/github/codeql-action/v2/node_modules/bottleneck/light.js:405:18) {
  status: 403,
  response: {
    url: 'https://api.github.com/repos/check-spelling-sandbox/codeql/code-scanning/analysis/status',
    status: 403,
```

codeql-action/init _currently_ requires a lot of stuff, that's currently https://github.com/github/codeql-action/pull/2110

I guess I should try to polish _that_ PR off before I finish _this_ PR.

I can test this branch (as `main` to deal w/ all the frustrating checks that only run on `main`) with my hacked version of the codeql action to see how this would fare...

File: .github/workflows/ql-for-ql-tests.yml
Comment (jsoref) [reply .github/workflows/ql-for-ql-tests.yml:22]: Without fixing:
* https://github.com/github/codeql-action/pull/2121

Things fail:
https://github.com/check-spelling-sandbox/codeql/actions/runs/7820458947

Using a fixed version of github/codeql-action, it won't fail:
https://github.com/check-spelling-sandbox/codeql/actions/runs/7824516168/job/21347184612

So, I'll be able to remove this once that PR merges (or considered differently, it's already removed locally, but I can't push that change to this branch until that PR merges).

File: .github/workflows/ruby-dataset-measure.yml
Comment (jsoref) [reply .github/workflows/ruby-dataset-measure.yml:22]: Indeed, it is not necessary: https://github.com/check-spelling-sandbox/codeql/actions/runs/7824516163/job/21347184556

File: .github/workflows/ruby-qltest.yml
Comment (jsoref) [reply .github/workflows/ruby-qltest.yml:34]: Maybe?

I can't get ruby to build (I think I need larger runners or to try to kill a bunch of processes to get my runs more memory):
https://github.com/check-spelling-sandbox/codeql/actions/runs/7820458922/job/21335248948

We'll see what things look like when I update this PR.

File: .github/workflows/swift.yml
Comment (jsoref) [reply .github/workflows/swift.yml:38]: Maybe?

This job consistently dies:
https://github.com/check-spelling-sandbox/codeql/actions/runs/7820458953/job/21335237762

We'll see what things look like when I update this PR.

But, I'd expect it to consistently die.

File: .github/workflows/validate-change-notes.yml
Comment (jsoref) [reply .github/workflows/validate-change-notes.yml:20]: Yes. https://github.com/check-spelling-sandbox/codeql/actions/runs/7820458929/job/21335231502

File: .github/workflows/csv-coverage-update.yml
Comment (jsoref) [reply .github/workflows/csv-coverage-update.yml:10]: Yes: https://github.com/check-spelling-sandbox/codeql/actions/runs/7824516174/job/21347184874

Comment (jsoref) [reply .github/workflows/csv-coverage-update.yml:10]: Actually, I'm not entirely certain what is necessary to make it generate a PR (maybe I need to borrow an older file?), but conceptually, that change is correct.

File: .github/workflows/query-list.yml
Comment (jsoref) [reply .github/workflows/query-list.yml:18]: Dropped

File: .github/workflows/check-implicit-this.yml
Comment (jsoref) [reply .github/workflows/check-implicit-this.yml:14]: With this pair of commits (and the various other things I've gotten merged), this now works: https://github.com/check-spelling-sandbox/codeql/actions/runs/7878541275

File: .github/workflows/ql-for-ql-build.yml
Comment (jsoref) [inline .github/workflows/ql-for-ql-build.yml:26]: @aeisenberg says:
> All of it really should be using `@main` since we want to test on the latest in case we break something.

**Review by aeisenberg — COMMENTED** [review]
Partial review. I'm finding it hard to verify that this PR is correct.

**Review by aeisenberg — APPROVED** [review]
This is looking good to me, but since this change affects lots of files, I'd like someone else to approve as well.

File: .github/workflows/csharp-qltest.yml
Comment (angelapwen) [inline .github/workflows/csharp-qltest.yml:58]: Do we need this if the permissions are already declared in top-level scope?

File: .github/workflows/fast-forward.yml
Comment (angelapwen) [inline .github/workflows/fast-forward.yml:12]: Can we remove the permissions block in the `fast-forward` job below as it's redundant?

**Review by angelapwen — COMMENTED** [review]
Some questions in comments. Also, I still see a bunch of `security-events: read` permissions here: are those necessary? I see that they were able to be dropped from a few workflows because of the changes you'd made in the Action 😄

**Review by aeisenberg — APPROVED** [review]

**Review by angelapwen — APPROVED** [review]

**Comment by angelapwen** [comment]
Thank you for your patience and of course contributions @jsoref!! Merging now 💕

**Pull request merged**
